### ISSUE TITLE
Full-FRM error: Py6S.sixs_exceptions.OutputParsingError: check for invalid parameter inputs

### ISSUE THREAD

**Issue by ARamsay17** [nasa/HyperCP#121]
When running full-FRM correction on my windows 10 installation of HCP I encounter this error when calculating the direct irradiance ratio.
- py6s version 1.9.2 installed via conda-forge.   
- conda version 23.7.2
- My hypercp environment was created using the provided .yml file packaged in the git repository.

Error message:

File "C:\Users\ar17\PycharmProjects\FRM4SOC-2-125922\HCP\Source\ProcessInstrumentUncertainties.py", line 1386, in FRM
    res_py6s = ProcessL1b_FRMCal.get_direct_irradiance_ratio(node, sensortype, called_L2=True)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ar17\PycharmProjects\FRM4SOC-2-125922\HCP\Source\ProcessL1b_FRMCal.py", line 97, in get_direct_irradiance_ratio
    wavelengths, res = Py6S.SixSHelpers.Wavelengths.run_wavelengths(s, 1e-3*wvl)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ar17\.conda\envs\hypercp\Lib\site-packages\Py6S\SixSHelpers\all_wavelengths.py", line 90, in run_wavelengths
    results = pool.map(f, wavelengths)
              ^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ar17\.conda\envs\hypercp\Lib\multiprocessing\pool.py", line 367, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ar17\.conda\envs\hypercp\Lib\multiprocessing\pool.py", line 774, in get
    raise self._value
  File "C:\Users\ar17\.conda\envs\hypercp\Lib\multiprocessing\pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
                    ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ar17\.conda\envs\hypercp\Lib\multiprocessing\pool.py", line 48, in mapstar
    return list(map(*args))
           ^^^^^^^^^^^^^^^^
  File "C:\Users\ar17\.conda\envs\hypercp\Lib\site-packages\Py6S\SixSHelpers\all_wavelengths.py", line 70, in f
    a.run()
  File "C:\Users\ar17\.conda\envs\hypercp\Lib\site-packages\Py6S\sixs.py", line 357, in run
    self.outputs = Outputs(outputs[0], outputs[1])
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ar17\.conda\envs\hypercp\Lib\site-packages\Py6S\outputs.py", line 70, in __init__
    raise OutputParsingError(
Py6S.sixs_exceptions.OutputParsingError: 6S returned an error (shown above) - check for invalid parameter inputs

**Comment by oceancolorcoder** [comment]
Workaround: reinstall py6S

**Comment by juanchossn** [comment]
Dirk also reported it in Linux

**Comment by robintw** [comment]
Hi - I'm the maintainer of Py6S, I was pointed to this issue by someone on the Py6S mailing list.

Looking at the error message above, there should be some other sort of error coming from the underlying 6S model directly. The final line says "6S returned an error (shown above)" - and somewhere above the stack trace there should be the raw output from the 6S model (it may appear on stderr not stdout, in case that affects how you find it).

If you're able to provide that error here then I can try and work out what's going on.

**Comment by juanchossn** [comment]
Thanks Robin for your quick reply! This error is not occurring regularly, but only occasionally, only for some of us, and a bit unpredictably for the moment. Once we run into it again we will look into detail in the stderr to see if we can provide you more detailed information.

**Comment by ARamsay17** [comment]
Hello Robin,

I got the error again today looking into the code it was seemingly caught by line 68 - if platform.system() != "Darwin":

Outputs:
platform.system() = 'Windows'
sys.version_info[0] = 3
stdout = b''
stderr =  b'The system cannot execute the specified program.\r\n'

**Comment by robintw** [comment]
This is interesting - it seems like the error is an error from the operating system saying that it can't run the 6S executable.

Is this definitely only happening intermittently? I'd understand it more if it were happening all the time, but an intermittent error like this is strange.

How did you install Py6S and the underlying 6S model? Was it using conda, or some other system?

If you know how to do this, then it'd be great to find the 6S executable and try running it from the command-line manually. That _should_ be as simple as opening the command prompt and cd'ing to the directory containing the 6S executable, and then just typing its name (probably `sixs.exe` on Windows). It won't do much (it'll hang waiting for input from an input file), but you should at least be able to tell if it gives that error straight away or if it runs at all. Happy to talk you through this if needed.

**Comment by robintw** [comment]
Ah, actually, I've just thought of another thing: it looks like you're running using `run_wavelengths` which runs things in parallel. It's possible that there's something strange going on with the parallelisation that is causing this error - I wouldn't expect it, but I've seen weirder things. Can you try turning parallelisation off for the `run_wavelengths` call and see if that stops the errors? You can do this by passing `n=1` to the `run_wavelengths` function (or any of the other functions you might use, like `run_vnir`).

**Comment by ARamsay17** [comment]
> Ah, actually, I've just thought of another thing: it looks like you're running using `run_wavelengths` which runs things in parallel. It's possible that there's something strange going on with the parallelisation that is causing this error - I wouldn't expect it, but I've seen weirder things. Can you try turning parallelisation off for the `run_wavelengths` call and see if that stops the errors? You can do this by passing `n=1` to the `run_wavelengths` function (or any of the other functions you might use, like `run_vnir`).

It's been a few months but I encountered the issue again. I attempted the fix you suggested and it worked. As a fix I will pass n=1 to run_wavelengths whenever the processor detects that it is being run on Windows. Many thanks for the suggestion.

**Comment by oceancolorcoder** [comment]
Looking back I never could find the commit that fixed this, but closing as resolved.
